ocaml/xapi/xapi_host.ml

```diff
@@ -1340,11 +1340,98 @@ let get_uncooperative_resident_VMs ~__context ~self = []
+  let open X509.Validation in
+  let now = Ptime_clock.now () in
+  match valid_ca ~time:now certificate with
+  | Ok () ->
+      ()
+  | Error e ->
+      let msg = Fmt.str "%a" pp_ca_error e in
+      D.error "assert_valid_ca: now='%s' msg='%s'" (Ptime.to_rfc3339 now) msg ;
+      raise
+        Api_errors.(Server_error (invalid_value, ["certificate"; "<omitted>"]))
+
+let get_host_certs ~__context ~host :
+    (API.ref_Certificate * API.certificate_t) list =
+  (* [get_host_certs ~__context ~host] gets all host certs in the database
+   * belonging to [host] (the term 'host' is overloaded here) *)
+  let open Db_filter_types in
+  let type' = Eq (Field "type", Literal "host") in
+  let host' = Eq (Field "host", Literal (Ref.string_of host)) in
+  let expr = And (type', host') in
+  Db.Certificate.get_records_where ~__context ~expr
+
+let get_ca_cert ~__context ~name : API.ref_Certificate * API.certificate_t =
+  let open Db_filter_types in
+  let type' = Eq (Field "type", Literal "ca") in
+  let name' = Eq (Field "type", Literal name) in
+  let expr = And (type', name') in
+  match Db.Certificate.get_records_where ~__context ~expr with
+  | [x] ->
+      x
+  | xs ->
+      D.error "get_ca_cert: found %i certificates with name='%s'. expected 1"
+        (List.length xs) name ;
+      raise
+        Api_errors.(Server_error (invalid_value, ["certificate:name"; name]))
+
+let add_cert_to_db ~__context ~host ~_type (certificate : X509.Certificate.t) :
+    API.ref_Certificate =
+  let name, _type =
+    (* only user installed CA certs have names *)
+    match _type with `host -> ("", `host) | `ca name -> (name, `ca)
+  in
+  let date_of_ptime time = Date.of_float (Ptime.to_float_s time) in
+  let dates_of_ptimes (a, b) = (date_of_ptime a, date_of_ptime b) in
+  let not_before, not_after =
+    dates_of_ptimes (X509.Certificate.validity certificate)
+  in
+  let fingerprint =
+    X509.Certificate.fingerprint Mirage_crypto.Hash.(`SHA256) certificate
+    |> Certificates.pp_hash
+  in
+  let uuid = Uuid.(to_string (make_uuid ())) in
+  let ref' = Ref.make () in
+  Db.Certificate.create ~__context ~ref:ref' ~uuid ~host ~not_before ~not_after
+    ~fingerprint ~name ~_type ;
+  ref'
+
+let uninstall_certificates ~__context refs : unit =
+  (* remove certs from the file system and the database *)
+  refs
+  |> List.iter @@ fun (self, record) ->
+     ( match record.API.certificate_type with
+     | `host ->
+         let name = record.API.certificate_uuid in
+         Certificates.(pool_uninstall CA_Certificate) ~__context ~name
+     | `ca ->
+         let name = record.API.certificate_name in
+         Certificates.(host_uninstall CA_Certificate name)
+     ) ;
+     Db.Certificate.destroy ~__context ~self
+
 let install_ca_certificate ~__context ~host ~name ~cert =
-  Certificates.(host_install CA_Certificate name cert)
+  let certificate = pem_of_string cert in
+  assert_valid_ca certificate ;
```

Comment:
Whilst not strictly necessary, I thought it would be nice to stop users accidentally installing a non-CA cert here

We could add a similar check that CA=false in the 'Basic Constraints' x509 extension, when we install a ```host`` cert

Reply:
I believe restricting the trusted certificates that users can install is out of scope for the feature, this should be part of a future enhancement.

In any case, right now self-signed certs without basic constraints are allowed to be installed so I'm wary to break currently-working configurations.

Reply:
Fair enough, I will remove this assertion